trio/_core/tests/test_run.py

```diff
@@ -2015,3 +2016,108 @@ def test_sniffio_integration():
     assert task.custom_sleep_data == 1
     await _core.checkpoint()
     assert task.custom_sleep_data is None
+
+
+@types.coroutine
+def async_yield(value):
+    yield value
+
+
+async def test_permanently_detach_coroutine_object():
+    task = None
+    pdco_outcome = None
+
+    async def detachable_coroutine(task_outcome, yield_value):
+        await sleep(0)
+        nonlocal task, pdco_outcome
+        task = _core.current_task()
+        pdco_outcome = await outcome.acapture(
+            _core.permanently_detach_coroutine_object, task_outcome
+        )
+        await async_yield(yield_value)
+
+    async with _core.open_nursery() as nursery:
+        nursery.start_soon(
+            detachable_coroutine, outcome.Value(None), "I'm free!"
+        )
+
+    # If we get here then Trio thinks the task has exited... but the coroutine
+    # is still iterable
+    assert pdco_outcome is None
+    assert task.coro.send("be free!") == "I'm free!"
+    assert pdco_outcome == outcome.Value("be free!")
+    with pytest.raises(StopIteration):
+        task.coro.send(None)
+
+    # Check the exception paths too
+    task = None
+    pdco_outcome = None
+    with pytest.raises(KeyError):
+        async with _core.open_nursery() as nursery:
+            nursery.start_soon(
+                detachable_coroutine, outcome.Error(KeyError()), "uh oh"
+            )
+    throw_in = ValueError()
+    assert task.coro.throw(throw_in) == "uh oh"
+    assert pdco_outcome == outcome.Error(throw_in)
+    with pytest.raises(StopIteration):
+        task.coro.send(None)
+
+    async def bad_detach():
+        async with _core.open_nursery():
+            with pytest.raises(RuntimeError) as excinfo:
+                await _core.permanently_detach_coroutine_object(
+                    outcome.Value(None)
+                )
+            assert "open nurser" in str(excinfo.value)
+
+    async with _core.open_nursery() as nursery:
+        nursery.start_soon(bad_detach)
+
+
+async def test_attach_and_reattach_coroutine_object():
+    unrelated_task = None
+    task = None
+
+    async def unrelated_coroutine():
+        nonlocal unrelated_task
+        unrelated_task = _core.current_task()
+
+    async def reattachable_coroutine():
+        await sleep(0)
+
+        nonlocal task
+        task = _core.current_task()
+
+        def abort_fn(_):  # pragma: no cover
```

Comment:
Might be worth adding a test that cancels a detached coroutine, to validate that the abort_fn does actually get called?

Reply:
Done